plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,192 @@ func haveSame(a1, a2 []string) bool {
+	checker := &PodAffinityChecker{
+		info:      info,
+		podLister: podLister,
+	}
+	return checker.InterPodAffinityMatches
+}
+
+func (checker *PodAffinityChecker) InterPodAffinityMatches(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
+	node, err := checker.info.GetNodeInfo(nodeName)
+	if err != nil {
+		return false, err
+	}
+	allPods, err := checker.podLister.List(labels.Everything())
+	if err != nil {
+		return false, err
+	}
+	return checker.NodeMatchPodAffinityAntiAffinity(pod, allPods, node), nil
+}
+
+// CheckIfAnyPodThatMatchPodAffinityTerm checks if any of given pods can match the specific podAffinityTerm.
+func (checker *PodAffinityChecker) CheckIfAnyPodThatMatchPodAffinityTerm(pod *api.Pod, allPods []*api.Pod, node *api.Node, podAffinityTerm api.PodAffinityTerm) (bool, error) {
+	for _, ep := range allPods {
+		match, err := priorityutil.CheckIfPodMatchPodAffinityTerm(ep, pod, podAffinityTerm,
+			func(ep *api.Pod) (*api.Node, error) { return checker.info.GetNodeInfo(ep.Spec.NodeName) },
+			func(pod *api.Pod) (*api.Node, error) { return node, nil },
+		)
+		if err != nil || match {
+			return match, err
+		}
+	}
+	return false, nil
+}
+
+// Checks whether the given node has pods which satisfy all the required pod affinity scheduling rules.
+// If node has pods which satisfy all the required pod affinity scheduling rules then return true.
+func (checker *PodAffinityChecker) CheckNodeMatchesHardPodAffinity(pod *api.Pod, allPods []*api.Pod, node *api.Node, podAffinity *api.PodAffinity) bool {
+	var podAffinityTerms []api.PodAffinityTerm
+	if len(podAffinity.RequiredDuringSchedulingIgnoredDuringExecution) != 0 {
+		podAffinityTerms = podAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+	}
+	// TODO: Uncomment this block when implement RequiredDuringSchedulingRequiredDuringExecution.
+	//if len(podAffinity.RequiredDuringSchedulingRequiredDuringExecution) != 0 {
+	//	podAffinityTerms = append(podAffinityTerms, podAffinity.RequiredDuringSchedulingRequiredDuringExecution...)
+	//}
+
+	for _, podAffinityTerm := range podAffinityTerms {
+		podAffinityTermMatches, err := checker.CheckIfAnyPodThatMatchPodAffinityTerm(pod, allPods, node, podAffinityTerm)
+		// if err or podAffinityTerm match failed then return false otherwise continue
+		if err != nil {
+			glog.V(10).Infof("Cannot schedule pod %+v onto node %v, an error ocurred when checking existing pods on the node for PodAffinityTerm %v err: %v",
+				podName(pod), node.Name, podAffinityTerm, err)
+			return false
+		}
+
+		if !podAffinityTermMatches {
+			// TODO: Think about whether this can be simplified once we have controllerRef
+			// Check if it is in special case that the requiredDuringScheduling affinity requirement can be disregarded.
+			// If the requiredDuringScheduling affinity requirement matches a pod's own labels and namespace, and there are no other such pods
+			// anywhere, then disregard the requirement.
```

Comment:
Add to the end of the comment: "This allows rules like "schedule all of the pods of this collection to the same zone" to not block forever because the first pod of the collection can't be scheduled."